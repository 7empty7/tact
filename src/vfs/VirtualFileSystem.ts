export type VirtualFileSystem = {
    resolve(...path: string[]): string;
    exists(path: string): boolean;
    readFile(path: string): Buffer;
    writeFile(path: string, data: Buffer | string): void;
};