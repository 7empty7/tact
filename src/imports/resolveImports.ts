import { parseImports } from '../grammar/grammar';
import { VirtualFileSystem } from '../vfs/VirtualFileSystem';
import { resolveLibrary } from './resolveLibrary';

export function resolveImports(args: { entrypoint: string, project: VirtualFileSystem, stdlib: VirtualFileSystem }) {

    //
    // Load stdlib and entrypoint
    //

    // const stdlibFuncPath = args.stdlib.resolve('./stdlib.fc');
    // const stdlibFunc = args.stdlib.readFile(stdlibFuncPath).toString();

    const stdlibTactPath = args.stdlib.resolve('./stdlib.tact');
    const stdlibTact = args.stdlib.readFile(stdlibTactPath).toString();

    const codePath = args.project.resolve(args.entrypoint);
    const code = args.project.readFile(codePath).toString();

    //
    // Resolve all imports
    //

    const importedTact: { code: string, path: string }[] = [];
    const importedFunc: { code: string, path: string }[] = [];
    const processed = new Set<string>();
    const pending: { code: string, path: string }[] = [];
    function processImports(source: string, path: string) {
        const imp = parseImports(source, path);
        for (const i of imp) {

            // Resolve library
            const resolved = resolveLibrary({
                path: path,
                name: i,
                project: args.project,
                stdlib: args.stdlib
            });
            if (!resolved.ok) {
                throw new Error(`Could not resolve import ${i} in ${path}`);
            }

            // Check if already imported
            if (resolved.kind === 'func') {
                if (importedFunc.find((v) => v.path === resolved.path)) {
                    continue;
                }
            } else {
                if (importedTact.find((v) => v.path === resolved.path)) {
                    continue;
                }
            }

            // Load code
            let vfs = resolved.source === 'project' ? args.project : args.stdlib;
            let code: string = vfs.readFile(resolved.path).toString();

            // Add to imports
            if (resolved.kind === 'func') {
                importedFunc.push({ code, path: resolved.path });
            } else {
                if (!processed.has(resolved.path)) {
                    processed.add(resolved.path);
                    pending.push({ path: resolved.path, code });
                }
            }
        }
    }

    // Run resolve
    importedTact.push({ code: stdlibTact, path: stdlibTactPath });
    processImports(stdlibTact, stdlibTactPath);
    processImports(code, codePath);
    while (pending.length > 0) {
        let p = pending.shift()!;
        importedTact.push(p);
        processImports(p.code, p.path);
    }
    importedTact.push({ code: code, path: codePath }); // To keep order same as before refactoring

    // Assemble result
    return {
        tact: [
            ...importedTact,
        ],
        func: [
            // { code: stdlibFunc, path: stdlibFuncPath },
            ...importedFunc
        ]
    };
}