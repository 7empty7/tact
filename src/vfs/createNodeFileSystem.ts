import { VirtualFileSystem } from "./VirtualFileSystem";
import fs from 'fs';
import path from "path";
import mkdirp from "mkdirp";

export function createNodeFileSystem(root: string, readonly: boolean = true): VirtualFileSystem {
    let normalizedRoot = path.normalize(root);
    return {
        root: normalizedRoot,
        exits(filePath: string): boolean {
            if (!filePath.startsWith(normalizedRoot)) {
                throw new Error('Path is outside of the root directory');
            }
            return fs.existsSync(filePath);
        },
        resolve(...filePath) {
            return path.normalize(path.resolve(normalizedRoot, ...filePath));
        },
        readFile(filePath) {
            if (!filePath.startsWith(normalizedRoot)) {
                throw new Error('Path is outside of the root directory');
            }
            return fs.readFileSync(filePath);
        },
        writeFile(filePath, content) {
            if (readonly) {
                throw new Error('File system is readonly');
            }
            if (!filePath.startsWith(normalizedRoot)) {
                throw new Error('Path is outside of the root directory');
            }

            mkdirp.sync(path.dirname(filePath));
            fs.writeFileSync(filePath, content);
        }
    }
}