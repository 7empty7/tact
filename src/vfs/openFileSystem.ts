import path from 'path';
import fs from 'fs';
import mkdirp from 'mkdirp';
import pathNormalize from 'path-normalize';
import { VirtualFileSystem } from './VirtualFileSystem';

export function openFileSystem(root: string): VirtualFileSystem {
    let rootNormalized = pathNormalize(root);
    return {
        resolve(...pathSegments: string[]): string {
            return pathNormalize(path.resolve(rootNormalized, ...pathSegments));
        },
        exists(filePath) {
            return fs.existsSync(filePath);
        },
        readFile(filePath) {
            return fs.readFileSync(filePath);
        },
        writeFile(filePath, data) {
            
            // Crete dirs if needed
            mkdirp.sync(path.dirname(filePath));

            // Write file
            fs.writeFileSync(filePath, data);
        },
    }
}