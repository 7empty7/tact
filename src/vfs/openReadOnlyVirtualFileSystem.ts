import normalize from "path-normalize";
import { VirtualFileSystem } from "./VirtualFileSystem";

export function openReadOnlyVirtualFileSystem(root: string, files: { [key: string]: string }): VirtualFileSystem {
    return {
        resolve(...pathSegments: string[]): string {
            return root + '/' + normalize(pathSegments.join('/'));
        },
        exists(filePath) {
            if (!filePath.startsWith(root + '/')) {
                return false;
            }
            return files.hasOwnProperty(filePath.slice(root.length + 1));
        },
        readFile(filePath) {
            if (!filePath.startsWith(root + '/')) {
                throw Error('Cannot read file outside of root');
            }
            if (!files.hasOwnProperty(filePath.slice(root.length + 1))) {
                throw Error('File does not exist');
            }
            return Buffer.from(files[filePath.slice(root.length + 1)], 'base64');
        },
        writeFile(filePath, data) {
            throw Error('Cannot write to read-only file system');
        }
    }
}