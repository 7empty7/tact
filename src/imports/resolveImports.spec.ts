import { resolveImports } from './resolveImports';
import { createNodeFileSystem } from '../vfs/createNodeFileSystem';
import path from 'path';

describe('resolveImports', () => {
    it('should resolve imports', () => {
        let project = createNodeFileSystem(path.resolve(__dirname, '__testdata', 'project'));
        let stdlib = createNodeFileSystem(path.resolve(__dirname, '__testdata', 'stdlib'));
        let resolved = resolveImports({
            project,
            stdlib,
            entrypoint: './main.tact'
        });
        expect(resolved).toMatchSnapshot();
    });
});