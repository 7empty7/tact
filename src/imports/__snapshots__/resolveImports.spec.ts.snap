// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveImports should resolve imports 1`] = `
{
  "func": [
    {
      "code": "",
      "path": "/Users/steve/Develop/tact/src/imports/__testdata/stdlib/stdlib2.fc",
    },
  ],
  "tact": [
    {
      "code": "import "./stdlib2.fc";",
      "path": "/Users/steve/Develop/tact/src/imports/__testdata/stdlib/stdlib.tact",
    },
    {
      "code": "",
      "path": "/Users/steve/Develop/tact/src/imports/__testdata/project/imported.tact",
    },
    {
      "code": "import "./imported";",
      "path": "/Users/steve/Develop/tact/src/imports/__testdata/project/main.tact",
    },
  ],
}
`;
